File: core/tactic_normalization.py
"""Pure tactic-data normalization. No filesystem, database, CLI, or print dependency."""
from __future__ import annotations

import copy
from typing import Any, Dict, List, Tuple


def normalize_tactic_data(tactic: Dict[str, Any], aliases: Dict[Tuple[str, str], str],
                          team_instruction_catalog: Dict[str, Any] | None = None) -> tuple[Dict[str, Any], List[Dict[str, str]]]:
    if not isinstance(tactic, dict):
        raise ValueError("Tactic must be a JSON object")
    normalized = copy.deepcopy(tactic)
    changes = []
    for phase, key in (("IP", "ip_roles"), ("OOP", "oop_roles")):
        roles = normalized.get(key)
        if not isinstance(roles, dict):
            continue
        for position, value in roles.items():
            replacement = aliases.get((phase, value), value) if isinstance(value, str) else value
            if replacement != value:
                roles[position] = replacement
                changes.append({"phase": phase, "position": str(position), "before": value, "after": replacement})
    if team_instruction_catalog is not None:
        from core.team_instructions import normalize_team_instructions
        for phase, key in (("IP", "ip_team_instructions"), ("OOP", "oop_team_instructions")):
            if key in normalized:
                normalized[key], _ = normalize_team_instructions(normalized[key], phase, team_instruction_catalog)
    return normalized, changes

File: web/static/js/tactic_normalization.js
export function normalizeTacticData(tactic, aliases, teamInstructionCatalog) {
  const normalized = structuredClone(tactic || {}), changes = [];
  for (const [phase, key] of [['IP', 'ip_roles'], ['OOP', 'oop_roles']]) {
    for (const [position, value] of Object.entries(normalized[key] || {})) {
      const replacement = typeof value === 'string' ? (aliases[phase]?.[value] || value) : value;
      if (replacement !== value) { normalized[key][position] = replacement; changes.push({phase, position, before:value, after:replacement}); }
    }
  }
  for (const [phase, key] of [['IP','ip_team_instructions'], ['OOP','oop_team_instructions']]) {
    if (key in normalized) normalized[key] = normalizeTeamInstructions(normalized[key], phase, teamInstructionCatalog);
  }
  return {normalized, changes};
}
export function normalizeTeamInstructions(raw, phase, catalog) {
  if (raw == null) return {};
  if (typeof raw !== 'object' || Array.isArray(raw)) throw new Error(`${phase.toLowerCase()}_team_instructions must be an object`);
  const byId = new Map((catalog.instructions || []).map(row => [row.internal_id,row]));
  const out = {};
  for (const [categoryId,valueId] of Object.entries(raw)) {
    const category=byId.get(categoryId);
    if (!category || category.phase !== phase) throw new Error(`Unknown or cross-phase team instruction: ${categoryId}`);
    const value=(category.selectable_values||[]).find(row=>row.internal_id===valueId);
    if (!value || value.verification !== 'user_ingame_verified') throw new Error(`Unverified team instruction value: ${valueId}`);
    out[categoryId]=valueId;
  }
  return out;
}
